lib/pack/aws/cache.ts

```diff
@@ -55,44 +57,56 @@ export class S3GoalCacheArchiveStore implements GoalCacheArchiveStore {
     public async store(gi: GoalInvocation, classifier: string, archivePath: string): Promise<string> {
         const file = fs.createReadStream(archivePath);
         return this.awsS3(
-            gi,
-            classifier,
-            async (storage, bucket, cachePath) =>
-                storage.putObject({ Bucket: bucket, Key: cachePath, Body: file }).promise(),
-            "store",
+          gi,
+          classifier,
+          async (storage, bucket, cachePath) =>
+            storage
+              .putObject({ Bucket: bucket, Key: cachePath, Body: file })
+              .promise(),
+          "store",
         );
     }
 
     public async delete(gi: GoalInvocation, classifier: string): Promise<void> {
         await this.awsS3(
-            gi,
-            classifier,
-            async (storage, bucket, cachePath) => storage.deleteObject({ Bucket: bucket, Key: cachePath }).promise(),
-            "delete",
+          gi,
+          classifier,
+          async (storage, bucket, cachePath) =>
+            storage.deleteObject({ Bucket: bucket, Key: cachePath }).promise(),
+          "delete",
         );
     }
 
     public async retrieve(gi: GoalInvocation, classifier: string, targetArchivePath: string): Promise<void> {
         await this.awsS3(
-            gi,
-            classifier,
-            async (storage, bucket, cachePath) => {
-                return new Promise(resolve => {
-                    storage
-                        .getObject({ Bucket: bucket, Key: cachePath })
-                        .createReadStream()
-                        .pipe(fs.createWriteStream(targetArchivePath))
-                        .on("close", () => resolve(targetArchivePath));
-                });
-            },
-            "retrieve",
+          gi,
+          classifier,
+          async (storage, bucket, cachePath) => {
+            return new Promise((resolve, reject) => {
+              storage
+                .getObject({ Bucket: bucket, Key: cachePath })
+                .createReadStream()
+                .on("error", reject)
+                .pipe(fs.createWriteStream(targetArchivePath))
+                .on("error", reject)
+                .on("end", () => resolve(targetArchivePath))
```

Comment:
Typically you should wait for the stream to close before resolving, but perhaps these  streams are different?